platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc

```diff
@@ -228,39 +403,44 @@ There are a few points to consider when using `registerTransform()`:
-For more information on how to use link:{groovyDslPath}/org.gradle.work.InputChanges.html[InputChanges], see the corresponding documentation for <<custom_tasks#incremental_tasks,incremental tasks>>.
+On the command line, Gradle runs tasks; not Artifact Transforms: `./gradlew build.`
+So how and when does it run transforms?
 
-Here is an example of an incremental transform that counts the lines of code in Java source files:
+A task can declare an input with attributes.
+When running this task, Gradle may run an Artifact Transform to ensure that the input has the proper attributes.
+Then, it will run the task.
 
-.Artifact transform for lines of code counting
-====
-include::sample[dir="snippets/dependencyManagement/artifactTransforms-incremental/kotlin",files="build.gradle.kts[tags=artifact-transform-countloc]"]
-include::sample[dir="snippets/dependencyManagement/artifactTransforms-incremental/groovy",files="build.gradle[tags=artifact-transform-countloc]"]
-====
-<1> Inject `InputChanges`
-<2> Query for changes in the input artifact
+There are two ways to execute a transform:
 
-// TODO: https://github.com/gradle/gradle/issues/10867
-// [[artifact-resolution-queries]]
-// == Artifact resolution queries
+1. Artifact Transforms execution for _project dependencies_ can be discovered ahead of task execution and, therefore can be scheduled outside the task execution (as a <<build_lifecycle.adoc#sec:execution, planned transform node in the execution graph>>).
+2. Artifact Transforms execution for _external module dependencies_ cannot be discovered ahead of task execution and, therefore are scheduled inside the task execution (as part of the <<build_lifecycle.adoc#sec:execution, task node in the execution graph>>).
+
+Unlike a task, which is invoked once all its dependencies are finished executing, many tasks can request transforms at any time in the build.
+Therefore, when an execution graph node or a task needs an artifact transform, it places a request.
+
+NOTE: Transform executions don’t run as ‘children’ of execution graph nodes. Requests are the ‘children’.
+
+Gradle will check if this request needs to be executed.
+The execution engine will fulfill the request as efficiently as possible.
+If the request is in the in-memory cache (daemon memory), `UP-TO-DATE`, or in the local/remove cache, Gradle will avoid executing it again.
+This means that one execution of a transform can fulfill many requests.
+Gradle takes full advantage of already fulfilled requests.
+
+TIP: In well-declared builds, project dependencies can be fully discovered during task configuration ahead of task execution scheduling.
+If the project dependency is badly declared (e.g., missing task input), the transform execution will happen inside the task.
+What could be a planned transform node in the execution graph becomes a transform request to the parent task.
```

Comment:
IMO much of this seems way to nitty gritty into the way the execution engine works. 

In practice I think the user really cares about a few things related to parallelism/caching:
- Artifact transforms can be run in parallel
- Artifact transforms are cached
- If separate resolutions used by different tasks require the same transform to be executed on the same artifacts, the transform results will be cached and shared. 

And then a few things related to _when_ transforms are executed:
- Transforms for project dependencies can be discovered early and can be scheduled to run before the task executes
- Transforms for external dependencies are not discovered early and will be executed as part of task execution
- Improper build dependency declarations can cause project artifact transforms to behave like external artifact transforms, delaying the transform to be executed as the task is executed.

Reply:
I wasn't sure how much we wanted to share in this section based on notes/PRs/discussions. I watered it down, but I'm not sure if it's to your liking yet. Let me know